client/lib/oauth-store/index.js

```diff
@@ -1,43 +1,53 @@
 /**
  * External dependencies
  */
-var debug = require( 'debug' )( 'calypso:auth:store' );
+import { get } from 'lodash/object';
+const debug = require( 'debug' )( 'calypso:auth:store' );
 
 /**
  * Internal dependencies
  */
 import { createReducerStore } from 'lib/store';
 import { actions as ActionTypes } from './constants';
-import { errors as errorTypes } from './constants';
+import { errors } from './constants';
 import * as OAuthToken from 'lib/oauth-token';
 
 /**
  * Module variables
  */
 const initialState = {
 	requires2fa: false,
+	pushauth: null,
 	inProgress: false,
 	errorLevel: false,
 	errorMessage: false
 };
 
 function handleAuthError( error, data ) {
-	let stateChanges = { errorLevel: 'is-error', requires2fa: false, inProgress: false };
-
-	stateChanges.errorMessage = data && data.body ? data.body.error_description : error.message;
+	const errorMessage = get( data, 'body.error_description', error.message );
+	const stateChanges = {
+		errorMessage,
+		errorLevel: 'is-error',
+		requires2fa: false,
+		inProgress: false
+	};
 
 	debug( 'Error processing login: ' + stateChanges.errorMessage );
 
-	if ( data && data.body ) {
-		if ( data.body.error === errorTypes.ERROR_REQUIRES_2FA ) {
-			stateChanges.requires2fa = true;
-			stateChanges.errorLevel = 'is-info';
-		} else if ( data.body.error === errorTypes.ERROR_INVALID_OTP ) {
-			stateChanges.requires2fa = true;
-		}
-	}
-
-	return stateChanges;
+	const errorCode = get( data, 'body.error', null );
+	const errorInfo = get( data, 'body.error_info', null );
+	const errorChanges = get( {
+		[ errors.ERROR_REQUIRES_2FA ]:
+			{ errorLevel: 'is-info', requires2fa: 'code' },
+		[ errors.ERROR_REQUIRES_2FA_PUSH_VERIFICATION ]:
+			{ errorLevel: 'is-info', requires2fa: 'push-verification', pushauth: errorInfo },
+		[ errors.ERROR_INVALID_PUSH_TOKEN ]:
+			{ errorLevel: 'is-info', requires2fa: 'push-verification' },
+		[ errors.ERROR_INVALID_OTP ]:
+			{ requires2fa: 'code' }
+	}, errorCode, {} );
```

Comment:
I remember seeing a `delete` statement here before, yet not any more. Do we need to unset a key somewhere? This can be effectively done by setting a key to `undefined` (though it's not 100% the same)